src/dparse/parser.d

```diff
@@ -3556,6 +3593,128 @@ class Parser
+     *     ;)
+     */
+    /*
+     * References:
+     * - [1] https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html
+     * - [2] https://wiki.dlang.org/Using_GDC
+     * - [3] https://github.com/dlang/dmd/blob/master/src/dmd/iasmgcc.d
+     *
+     * Separated into a different method because one cannot interleave DMD & GCC asm
+     * <asm-qualifiers> (volatile, inline, goto) not supperted (yet?)
+     */
+    GccAsmInstruction parseGccAsmInstruction()
+    {
+        mixin(traceEnterAndExit!(__FUNCTION__));
+        const startIndex = index;
+        auto node = allocator.make!GccAsmInstruction();
+
+        // Allow empty asm instructions
+        if (currentIs(tok!";"))
+        {
+            warn("Empty asm instruction");
+            node.tokens = tokens[startIndex .. index];
+            return node;
+        }
+
+        mixin(parseNodeQ!("node.assemblerTemplate", "Expression"));
+
+        // GDC allows e.g. asm { mixin(<some asm instruction>); }
+        if (!currentIs(tok!";"))
+        {
+            mixin(tokenCheck!":");
+            mixin(parseNodeQ!(`node.outputOperands`, `GccAsmOperandList`));
+
+            if (skip(tok!":"))
+            {
+                mixin(parseNodeQ!(`node.inputOperands`, `GccAsmOperandList`));
+
+                if (skip(tok!":"))
+                {
+                    if (!currentIs(tok!":"))
+                        mixin(parseNodeQ!("node.registers", "StringLiteralList"));
+
+                    if (skip(tok!":"))
+                    {
+                        if (node.outputOperands.items.length)
+                            error("goto-labels only allowed without output operands!", false);
+
+                        mixin(parseNodeQ!("node.gotos", "DeclaratorIdentifierList"));
+                    }
+                }
+            }
+        }
+
+        node.tokens = tokens[startIndex .. index];
+        return node;
+    }
+
+    /**
+     * Parses a GccAsmOperandList
+     *
+     * $(GRAMMAR $(RULEDEF gccAsmOperandList)
+     *     ($(RULE gccAsmOperand) ($(LITERAL ',') $(RULE gccAsmOperand))* )?
+     * )
+     */
+    GccAsmOperandList parseGccAsmOperandList()
+    {
+        mixin(traceEnterAndExit!(__FUNCTION__));
+
+        // Disambiguate between an empty and a missing list
+        if (currentIsOneOf(tok!":", tok!";"))
+            return allocator.make!GccAsmOperandList();
+
+        return parseCommaSeparatedRule!(GccAsmOperandList, GccAsmOperand)();
+    }
+
+    /**
+     * Parses a GccAsmOperand
+     *
+     * $(GRAMMAR $(RULEDEF gccAsmOperand)
+     *     ($(LITERAL '[') $(RULE identifier) $(LITERAL ']'))? $(RULE stringLiteral) $(LITERAL '(') $(RULE assignExpression) $(LITERAL ')')
+     * )
+     */
+    GccAsmOperand parseGccAsmOperand()
+    {
+        mixin(traceEnterAndExit!(__FUNCTION__));
+        if (currentIsOneOf(tok!":", tok!";"))
+            return null;
```

Comment:
Normally returning null indicates a parse error.

Reply:
This method is called from `parseCommaSeparatedRule` which stops when this rule returns `null`. Or is there  a better way to do this?

Reply:
I think that function works by parsing one of the comma-separated things, then checking to see if the current token is a comma, if it's not, then it stops parsing elements. It shouldn't be necessary for this to return null in the case that the list is done, because parseCommaSeparatedRule shouldn't try to call this function.

Reply:
You're right, that wasn't necessary anymore. Removed.